File: Array/6_unionArrays.cpp
// ramnitcode27
// Given two arrays a[] and b[] of size n and m respectively. The task is to find union between these two arrays.
// Union of the two arrays can be defined as the set containing distinct elements from both the arrays. If there are repetitions, then only one occurrence of element should be printed in the union.



#include <bits/stdc++.h>
using namespace std;

class Solution{
    public:
	// Time Complexity: O(mlog(m)+nlog(n))
    int doUnion(int a[], int n, int b[], int m)  {
        set<int> s;
        for (int i=0; i<n; i++) s.insert(a[i]);
        for (int i=0; i<m; i++) s.insert(b[i]);
        return s.size();
    }

	// Time Complexity: O(nlog(n))
	void intersection(int a[], int b[], int n, int m){
		sort(a, a+n);
		sort(b, b+m);
    	int i = 0, j = 0;
    	while(i < n && j < m) {
        	if(a[i] > b[j]) j++;
        	else if(b[j] > a[i]) i++;
        	else {
            	cout << a[i] << " ";
            	i++;
            	j++;
        	}
    	}
	}
};

int main() {
	
	int t;
	cin >> t;
	
	while(t--){
	    
	    int n, m;
	    cin >> n >> m;
	    int a[n], b[m];
	   
	    for(int i = 0;i<n;i++)
	       cin >> a[i];
	       
	    for(int i = 0;i<m;i++)
	       cin >> b[i];
	    Solution ob;
	    cout << ob.doUnion(a, n, b, m) << endl;
	    
	}
	
	return 0;
}

File: Array/6_unionArrays.py
# ramnitcode27
# Given two arrays a[] and b[] of size n and m respectively. The task is to find union between these two arrays.
# Union of the two arrays can be defined as the set containing distinct elements from both the arrays. If there are repetitions, then only one occurrence of element should be printed in the union.

class Solution:    
    # Time Complexity: O(m+n)
    # O(m+n) in case of Python because in python the set built-in method is quite different than that of C++ once, Python uses an hash map internally.
    def doUnion(self,a,n,b,m):
        # c=a+b
        # c=set(c)
        return len(set(a+b))
    
    # Time Complexity: O(nlogn)
    def intersection(a, b, n, m):
        i = 0
        j = 0
        # sort
        a.sort()
        b.sort()
        while (i < n and j < m):
            if (a[i] > b[j]):
                j += 1
                
            elif b[j] > a[i]:
                i += 1
            else:
                print(a[i], end=" ")
                i += 1
                j += 1

if __name__=='__main__':
    t=int(input())
    for _ in range(t):
        
        n,m=[int(x) for x in input().strip().split()]
        
        a=[int(x) for x in input().strip().split()]
        b=[int(x) for x in input().strip().split()]
        ob=Solution()
        
        print(ob.doUnion(a,n,b,m))